packages/prelude/src/array.ts

```diff
@@ -29,6 +29,23 @@ export function append <A>(x: A, a: A[]): A[] {
   return b
 }
 
+/**
+ * Concats two `ArrayLike`s
+ */
+export function concat <A>(a: ArrayLike<A>, b: ArrayLike<A>): A[] {
```

Comment:
Accepting `ArrayLike` in all the prelude array functions is great 👍.  Can we also return `ReadonlyArray` from all of them?

Reply:
Not sure that I see the reason - why should covariant return type be the smallest possible?

Reply:
I think that the return type should be restricted to an immutable structure. TS 3.4 introduced the syntax `readonly A[]`. See https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#improvements-for-readonlyarray-and-readonly-tuples

Reply:
> why should covariant return type be the smallest possible?

Because an ordinary array is mutable and @most/prelude has a focus on immutability.

Reply:
Ok, I tried. If I change `mapArray`'s return type to `ReadonlyArray<A>` I end up with a bunch of weird errors in `invoke` and as a consequence in `combine`, `zip`, etc.

Reply:
Hmmm.  I'm sure there's a way to write a type for mapArray that supports heterogenous tuples, but I don't feel like we need to go down that road right now.

Is there a way to add a few strategically placed type casts/assertions to make it work?  If so, I'd be ok with that.

Reply:
Done

Reply:
Aaaaand now https://github.com/mostjs/core/pull/296/files/253909b619d896f3423a6d97efcf86378d56addc..64293909ff6a7eb1f7a19d5b0374f2086aadb160#diff-bf49420dee5ae6f33cc0598cf149d325R142 is broken because Timeline mutates `timeslot.events`.

Reply:
> Aaaaand now https://github.com/mostjs/core/pull/296/files/253909b619d896f3423a6d97efcf86378d56addc..64293909ff6a7eb1f7a19d5b0374f2086aadb160#diff-bf49420dee5ae6f33cc0598cf149d325R142 is broken because Timeline mutates `timeslot.events`.

Ouch! I guess we need to look closer at when mutation should be allowed.

Reply:
@briancavalier ?

Reply:
Historically, there are internal places in `most/core` where we used array mutation for performance reasons.  In general, I don't feel like we should try to refactor all those places in this PR.  I'm totally ok with adding type casts to get this merged, and then addressing those later if/when we feel it's important.

IMHO, there is more benefit in moving forward with a great-yet-imperfect initial TS port, than trying to be perfect on first go.

In the particular case of Timeline `timeslot.events`, a typecast [inside `removeAllFrom`](https://github.com/mostjs/core/pull/296/files/253909b619d896f3423a6d97efcf86378d56addc..64293909ff6a7eb1f7a19d5b0374f2086aadb160#diff-bf49420dee5ae6f33cc0598cf149d325R142) seems very isolated and reasonable.  I vote for that :)

If that seems too aggressive for this PR, then I'm fine with using read-write array return types in prelude, and making them `readonly` in a future PR.

So, my preference on priority would be:

1. If possible and a reasonably small blast radius (subjective, but I trust your judgement, @raveclassic), return `ReadonlyArray`s and use isolated typecasts where necessary.
2. If not, return read-write arrays.